runtime/recipe/type-checker.js

```diff
@@ -37,102 +40,56 @@ class TypeChecker {
     return Type.newInterface(shape);
   }
 
-  static _coerceTypes(left, right) {
-    let leftType = left.type;
-    let rightType = right.type;
-
-    assert(!leftType.hasVariableReference);
-    assert(!rightType.hasVariableReference);
+  // left, right: {type, direction, connection}
+  static compareTypes(left, right) {
+    let [leftType, rightType] = Type.unwrapPair(left.type, right.type);
 
-    while (leftType.isSetView && rightType.isSetView) {
-      leftType = leftType.primitiveType();
-      rightType = rightType.primitiveType();
+    // TODO: we need a generic way to evaluate type compatibility
+    //       shapes + entities + etc
+    if (leftType.isInterface && rightType.isInterface) {
+      if (leftType.interfaceShape.equals(rightType.interfaceShape)) {
+        return {type: left, valid: true};
+      }
     }
 
-    leftType = leftType.resolvedType();
-    rightType = rightType.resolvedType();
-
-    if (leftType.equals(rightType))
-      return left;
-
-    // TODO: direction?
-    let type;
-    if (leftType.isVariable) {
-      leftType.variable.resolution = rightType;
-      type = right;
-    } else if (rightType.isVariable) {
-      rightType.variable.resolution = leftType;
-      type = left;
-    } else {
-      return null;
+    if (!leftType.isEntity || !rightType.isEntity) {
+      // TODO: direction?
+      if (leftType.isVariable) {
+        return {type: right, valid: true};
+      } else if (rightType.isVariable) {
+        return {type: left, valid: true};
```

Comment:
Need to establish one as the resolution of the other here.
leftType.variable.resolution = rightType;
or opposite.

Reply:
This now happens at the end of `processTypeList`: `item.type.resolveTo(baseType.type);`